core/libcamera_app.cpp

```diff
@@ -144,13 +171,71 @@ void LibcameraApp::CloseCamera()
 		LOG(2, "Camera closed");
 }
 
+Mode LibcameraApp::selectModeForFramerate(const libcamera::Size req, double fps)
+{
+	auto scoreFormat = [](double desired, double actual) -> double
+	{
+		double score = desired - actual;
+		// Smaller desired dimensions are preferred.
+		if (score < 0.0)
+			score = (-score) / 8;
+		// Penalise non-exact matches.
+		if (actual != desired)
+			score *= 2;
+
+		return score;
+	};
+
+	constexpr float penalty_AR = 1500.0;
+	constexpr float penalty_BD = 500.0;
+	constexpr float penalty_FPS = 2000.0;
+
+	double best_score = std::numeric_limits<double>::max(), score;
+	SensorMode best_mode;
+
+	LOG(1, "Mode selection:");
+	for (const auto &mode : sensor_modes_)
+	{
+		double reqAr = static_cast<double>(req.width) / req.height;
+		double fmtAr = static_cast<double>(mode.size.width) / mode.size.height;
+
+		// This is a really ugly way of getting the bit depth of the format.
+		// But apart from duplicating the massive bayer format table, there is
+		// no other way to determine this.
+		std::string fmt = mode.format.toString();
+		unsigned int mode_depth = fmt.find("8") != std::string::npos ? 8 :
+								  fmt.find("10") != std::string::npos ? 10 :
+								  fmt.find("12") != std::string::npos ? 12 : 16;
```

Comment:
Would it be tidier to hide this in a `depth()` method of the `SensorMode`? Again, don't really mind...

Reply:
Yes, I can do that!